File: panel/frontend/src/inertia/sharedProps.ts
import {usePage} from '@inertiajs/react'

/**
 * Field name to message, as written by `InertiaFlash.errors(...)` on the server. The
 * keys are the `name` attributes of the inputs that produced them, so a form can render
 * each message under the right control without a translation table.
 */
export type FieldErrors = Record<string, string>

/** One-off notices that survive exactly one redirect. */
export type FlashMessages = {
  success?: string
  error?: string
}

/**
 * The props every page receives whatever its controller put in the model.
 *
 * `errors` and `flash` come from `SharedProps.java` and are always objects, never
 * undefined - a page that has to write `errors?.name` in one place and `errors.name` in
 * another eventually gets it wrong where nobody looked. Anything a
 * `SharedPropsContributor` adds on the server belongs in this type too, so a page that
 * reads it gets a compile error when the server stops sending it.
 */
export interface SharedProps {
  errors: FieldErrors
  flash: FlashMessages
  [key: string]: unknown
}

/** The shared props, typed. Page-specific props come from `usePage<YourProps>()`. */
export function useSharedProps(): SharedProps {
  return usePage().props as unknown as SharedProps
}

/** The message for one field, or undefined when that field is fine. */
export function useFieldError(field: string): string | undefined {
  return useSharedProps().errors[field]
}

File: panel/src/main/java/lhqm/furimeo/wisper/web/SharedProps.java
package lhqm.furimeo.wisper.web;

import java.util.LinkedHashMap;
import java.util.Map;

import org.springframework.beans.factory.ObjectProvider;
import org.springframework.web.servlet.support.RequestContextUtils;

import jakarta.servlet.http.HttpServletRequest;

/**
 * Collects the props every page gets: validation errors and flash messages carried
 * across a redirect, plus whatever the {@link SharedPropsContributor} beans add.
 *
 * <p>{@code errors} and {@code flash} are always present, even when empty. A page that
 * has to write {@code props.errors?.name} in one place and {@code props.errors.name} in
 * another eventually gets it wrong in the place nobody tested.
 */
public class SharedProps {

    /** Flash attribute holding field name to message, written by {@link InertiaFlash}. */
    static final String ERRORS = "errors";

    /** Flash attribute holding one-off notices, written by {@link InertiaFlash}. */
    static final String FLASH = "flash";

    private final ObjectProvider<SharedPropsContributor> contributors;

    public SharedProps(ObjectProvider<SharedPropsContributor> contributors) {
        this.contributors = contributors;
    }

    Map<String, Object> forRequest(HttpServletRequest request) {
        Map<String, Object> props = new LinkedHashMap<>();
        props.put(ERRORS, flashValue(request, ERRORS));
        props.put(FLASH, flashValue(request, FLASH));
        contributors.orderedStream().forEach(contributor -> contributor.contribute(props, request));
        return props;
    }

    /**
     * Reads one flash attribute written by the request that redirected here.
     *
     * <p>Spring's flash map survives exactly one redirect, which is the lifetime a
     * "saved" banner should have. An empty map rather than null so the client can index
     * into it unconditionally.
     */
    private static Map<String, Object> flashValue(HttpServletRequest request, String key) {
        Map<String, ?> input = RequestContextUtils.getInputFlashMap(request);
        if (input == null) {
            return Map.of();
        }
        Object value = input.get(key);
        if (!(value instanceof Map<?, ?> map)) {
            return Map.of();
        }
        Map<String, Object> copy = new LinkedHashMap<>();
        map.forEach((name, message) -> copy.put(String.valueOf(name), message));
        return copy;
    }
}
